### DISCUSSION TITLE
Possible to replace or add BLE characteristics in default services?

### DISCUSSION THREAD

**Discussion by ajchellew** [micropython/micropython#12823] [RP2040 / Pico] [answered]
I am trying to emulate a device using MicroPython, I'm not entirely sure if this is specific to the Pico or the MicroPython implementation.

If I register Generic Access and Generic Attribute Services they appear in addition to those that are provided by default. 

![image](https://github.com/micropython/micropython/assets/17216760/6b30ff81-51c7-4939-be25-82a40bd9c2ac)

The problem is the connecting device requires the appearance `0x2A01` characteristic (amongst others) to be set and finds the default service first, which doesn't match what is expected.

Is it possible either to disable the default services or change characteristics in them? Besides `ble.config(gap_name=device_name)`

Also if I add the 5 services the device requires, only 4 of them (in addition to the existing 2 default ones) appear in nRF. Not sure if that's a BLE thing or a bug but thought it was worth mentioning.

**Comment by jimmo** [comment, marked as answer]
@ajchellew Unfortunately this is not possible right now.

I'm not sure what the best option is for how to provide this:
a) A way to disable the default services, which means the user is expected to provide them explicitly
b) Make it possible to configure all the options.

I'm leaning towards (b) because there are a few things in the default services that are a bit of a pain to implement by hand.

If you're able to build your own firmware, it's easy enough to disable the built-in default gap & gatt services.
For nimble: https://github.com/micropython/micropython/blob/master/extmod/nimble/modbluetooth_nimble.c#L653 & https://github.com/micropython/micropython/blob/master/extmod/nimble/modbluetooth_nimble.c#L920
For btstack: https://github.com/micropython/micropython/blob/master/extmod/btstack/modbluetooth_btstack.c#L890

**Reply by ajchellew** [reply]
Thanks @jimmo I expected as much. The pointers to the firmware will be useful, I don't yet know but I only got a Pico and started using MicroPython 2 days ago.

Whilst I think my use case is odd, I assume to make a HID thing you'd need to be able to add appearance into the default service. So there is hopefully some merit in it.

**Comment by xyzzy42** [comment]
I added an option for `gap_appearance`.  There's already an option for name.  And the NimBLE stack itself already supports configuring the appearance of the built in GAP service.

Option A seems wasteful, as one would need to duplicate all the code to setup the GAP service just to be able to change one byte from the default value.
